qdax/core/emitters/cma_emitter.py

```diff
@@ -0,0 +1,374 @@
+        """
+
+        # retrieve elements from the emitter state
+        cmaes_state = emitter_state.cmaes_state
+
+        # Compute the improvements - needed for re-init condition
+        indices = get_cells_indices(descriptors, repertoire.centroids)
+        improvements = fitnesses - emitter_state.previous_fitnesses[indices]
+
+        ranking_criteria = self._ranking_criteria(
+            emitter_state=emitter_state,
+            repertoire=repertoire,
+            genotypes=genotypes,
+            fitnesses=fitnesses,
+            descriptors=descriptors,
+            extra_scores=extra_scores,
+            improvements=improvements,
+        )
+
+        # get the indices
+        sorted_indices = jnp.flip(jnp.argsort(ranking_criteria))
+
+        # sort the candidates
+        sorted_candidates = jax.tree_util.tree_map(
+            lambda x: x[sorted_indices], genotypes
+        )
+        sorted_improvements = improvements[sorted_indices]
+
+        # compute reinitialize condition
+        emit_count = emitter_state.emit_count + 1
+
+        # check if the criteria are too similar
+        sorted_criteria = ranking_criteria[sorted_indices]
+        flat_criteria_condition = (
+            jnp.linalg.norm(sorted_criteria[0] - sorted_criteria[-1]) < 1e-12
+        )
+
+        # check all conditions
+        reinitialize = (
+            jnp.all(improvements < 0) * (emit_count > self._min_count)
+            + (emit_count > self._max_count)
+            + self._cmaes.stop_condition(cmaes_state)
+            + flat_criteria_condition
+        )
+
+        # If true, draw randomly and re-initialize parameters
+        def update_and_reinit(
+            operand: Tuple[
+                CMAESState, CMAEmitterState, MapElitesRepertoire, int, RNGKey
+            ],
+        ) -> Tuple[CMAEmitterState, RNGKey]:
+            return self._update_and_init_emitter_state(*operand)
+
+        def update_wo_reinit(
+            operand: Tuple[
+                CMAESState, CMAEmitterState, MapElitesRepertoire, int, RNGKey
+            ],
+        ) -> Tuple[CMAEmitterState, RNGKey]:
+            """Update the emitter when no reinit event happened.
+
+            Here lies a divergence compared to the original implementation. We
+            are getting better results when using no mask and doing the update
+            with the whole batch of individuals rather than keeping only the one
+            than were added to the archive.
+
+            Interestingly, keeping the best half was doing better. We think that
+            this might be due to the small batch size used.
+
+            This applies for the setting from the paper CMA-ME. Those facts might
+            not be true with other problems and hyperparameters.
+            """
+
+            (cmaes_state, emitter_state, repertoire, emit_count, random_key) = operand
+
+            # Update CMA Parameters
+            # mask = sorted_improvements >= 0
+            # mask = mask + 1e-6
```

Comment:
Why are those here? If they come from the original implementation, maybe say it in the comment on the previous line?

Reply:
I moved them to the docstrings with an explanation!